### PR TITLE
Pull Gym env in favor of Gymnasium and include ROMs with PyPI

### PR DESCRIPTION
**Pull request by pseudo-rnd-thoughts**
With Gymnasium v1.0 being released soon, one of the features removed is the namespace entry point used to register environments behind the scene, through what is a hack imho. 
In short, previous with `ale-py` (and `shimmy`) installed, the following code is possible
```python
import gymnasium as gym

env = gym.make("ALE/Pong-v5")
```
That despite `ale-py` not being installed, then behind the scenes, Atari environments are still registered. 

While an interesting approach, the more this has been used, the more of a headache it has been found to be. 
Therefore, it was decided to be removed in v1.0

This means that future users would be required to do
```python
import gymnasium as gym
import ale_py

env = gym.make("ALE/Pong-v5")
```
explicitly, importing ale_py for the modules to be loaded

As a result, this PR makes the following changes
* remove gym env
* Add gymnasium env 
* Includes AutoROM roms within `pypi` (but not github) 
* Add python 3.12 support

### CODE REVIEW

**Comment by pseudo-rnd-thoughts** [comment]
@JesseFarebro Hey, this is holding up Gymnasium's alpha 2, could you review this as soon as possible. Otherwise, I'll need to merge in the next week or so

File: .github/workflows/ci.yml
Comment (JesseFarebro) [inline .github/workflows/ci.yml:65]: Any reason CI is disabled for Windows?

File: .github/workflows/release.yml
Comment (JesseFarebro) [inline .github/workflows/release.yml:21]: Could use a more descriptive name.

File: .pre-commit-config.yaml
Comment (JesseFarebro) [inline .pre-commit-config.yaml:30]: Should switch to ruff?

File: .pre-commit-config.yaml
Comment (JesseFarebro) [inline .pre-commit-config.yaml:43]: Should switch to ruff?

File: .pre-commit-config.yaml
Comment (JesseFarebro) [inline .pre-commit-config.yaml:38]: Should switch to ruff?

File: .pre-commit-config.yaml
Comment (JesseFarebro) [inline .pre-commit-config.yaml:48]: Should switch to ruff?

File: curl_tar_gz_b64.sh
Comment (JesseFarebro) [inline curl_tar_gz_b64.sh:1]: Would be nice to have a more descriptive name. Also, since this is CI related would be nice to put it in .github/scripts or something.

File: md5.txt
Comment (JesseFarebro) [inline md5.txt:1]: Why is this being deleted? I would like to keep this as this is what we should abide by and I want to keep the history of this file for forensic purposes.

File: pyproject.toml
Comment (JesseFarebro) [inline pyproject.toml:13]: Maybe 3.9 as 3.8 is going to be EOL before the next major release?

File: pyproject.toml
Comment (JesseFarebro) [inline pyproject.toml:59]: Why include the tar.gz instead of just the bin files? I don't really like this.

**Review by JesseFarebro — CHANGES_REQUESTED** [review]
Overall this looks fine, my biggest complaint is around getting rid of the ROM plugins. Just assuming that the ROMs will be packaged and this will work for everyone's environment isn't going to cut it. The ALE is widely used at many companies where this isn't true so having a more customizable way to load ROMs is required IMO. I don't see any reason why you can't do the following:

- Package the bin files during CD
- Use the ROM plugin infrastructure to load these ROMs from `ale_py.roms`, in fact, it already did this, there should be no need to change anything here. All that needed to be done is to extract the ROMs into `ale_py.roms` during the wheel build.

File: .github/workflows/release.yml
Comment (pseudo-rnd-thoughts) [reply .github/workflows/release.yml:21]: The CI is broken for an unknown reason, see https://github.com/Farama-Foundation/Arcade-Learning-Environment/issues/515 for more info

File: .pre-commit-config.yaml
Comment (pseudo-rnd-thoughts) [reply .pre-commit-config.yaml:30]: We just copied in Gymansium's pre-commit, it certainly would be an option

File: md5.txt
Comment (pseudo-rnd-thoughts) [reply md5.txt:1]: @jjshoots Why did we change to `md5.txt` to `md5.json`?

File: pyproject.toml
Comment (pseudo-rnd-thoughts) [reply pyproject.toml:59]: @jjshoots You preferred to include the tar and not the bins, do you remember why?

File: src/python/env.py
Comment (pseudo-rnd-thoughts) [reply src/python/env.py:158]: As the class inherits from `gym.Env` which defines the `action_space` as an attribute and not a property then it feels strange to use a property here.

File: src/python/roms/__init__.py
Comment (pseudo-rnd-thoughts) [reply src/python/roms/__init__.py:24]: Thanks for information, that is very helpful to know. We will look to see what we can do. 
If the ROMs are packaged inside the PyPI install, does that cause issues for google? 
The primarily issue I see is for users with additional external ROMs not included in AutoROMs standard set then I'm not sure how this would work with it.

File: pyproject.toml
Comment (jjshoots) [reply pyproject.toml:59]: @pseudo-rnd-thoughts Someone along the chain of command mentioned we cannot have the bins on the GH repo.

**Comment by pseudo-rnd-thoughts** [comment]
Thanks for the review @JesseFarebro, I will talk to @jjshoots about the changes to the plugin system. 
It might be helpful to have a short discord call about this, if possible, to find to better route through. 
You can easily find me on the Gymnasium discord

The other suggested changes about Ruff, we can easily look at and could you look at the Windows CI issues linked above

**Comment by jjshoots** [comment]
Thanks for the review @JesseFarebro. I'm addressing the inclusion of bins on CD at #520.

File: .github/workflows/ci.yml
Comment (pseudo-rnd-thoughts) [reply .github/workflows/ci.yml:65]: See https://github.com/Farama-Foundation/Arcade-Learning-Environment/issues/515, we commented it out such that the CI passed in the meantime

File: src/python/roms/__init__.py
Comment (pseudo-rnd-thoughts) [reply src/python/roms/__init__.py:15]: Could you elaborate "history purposes"? What history data is of interest?

File: src/python/roms/__init__.py
Comment (pseudo-rnd-thoughts) [reply src/python/roms/__init__.py]: Done

**Comment by pseudo-rnd-thoughts** [comment]
@JesseFarebro Could you re-review? 

* changing to Ruff - We will investigate this but for simplicity, for this version, we will keep the updated pre-commit settings
* md5.txt - I don't understand the point about history, looking at the blame, there seem to be three different PRs that have updated it, none of them look critical to keep. Therefore, it seems to me reasonable to keep the updated JSON format. If I've missed something please say
*  plugin system - With the ROMs being included within the project downloads (except git clone), to my understanding, the only feature missing from the updated method is interoperability with atari-py roms. What feature in the plugin system that the updated version doesn't include is important for organisations, i.e., google. Otherwise, the updated version is much simpler for everyone, therefore, I would wish to remove it.

File: src/python/roms/__init__.py
Comment (JesseFarebro) [inline src/python/roms/__init__.py:27]: Maybe prefer to use `importlib.resources` which will provide better compatibility.

**Review by JesseFarebro — COMMENTED** [review]
Overall, looks good. Please consider my comment RE: ROM loading via an environment variable and if that's fixed I don't have any further comments.

---

> md5.txt - I don't understand the point about history, looking at the blame, there seem to be three different PRs that have updated it, none of them look critical to keep. Therefore, it seems to me reasonable to keep the updated JSON format. If I've missed something please say

It's fine, I can go hunting in the history to find any changes to `md5.txt`.

> plugin system - With the ROMs being included within the project downloads (except git clone), to my understanding, the only feature missing from the updated method is interoperability with atari-py roms. What feature in the plugin system that the updated version doesn't include is important for organisations, i.e., google. Otherwise, the updated version is much simpler for everyone, therefore, I would wish to remove it.

I took a look at the latest changes and this still wouldn't work for Google at least. At minimum, we would need a way to specify a different or additional directory to search for ROMs. If you really don't want to bring back the ROM plugins could you provide this functionality via an environment variable?

**Comment by jjshoots** [comment]
@JesseFarebro Apologies for seemingly going in circles with this. What component of the plugin system is required at Google?

More specifically, the current release already includes the binaries as part of package data on PyPI. Running `pip3 install ale-py` would pull the binaries from the python index itself at build, it is no longer "lazy-downloaded" at runtime.

The only cases I can see where the plugin system is needed are:
1. Google scrubs all `.bin` file from packages downloaded via pip.
2. There are some atari games not within ale-py that Google is using, and requires external linking
3. There are modifications to existing atari bins that are not native to ale-py.

Either 3 of those are valid use cases, but it'd be helpful to know which one in specific Google (or other companies) are using so that the plugin system's inclusion is justified.

Once again, sorry if this is going around in circles, we just want to make sure that the package doesn't have code that non of the maintainers know why should exist.

**Comment by JesseFarebro** [comment]
@jjshoots Google doesn't use pip, only the source code exists in our version control system. This is why we need a way at runtime to modify the path where the ROMs can be discovered. As I said in my previous comment, at minimum we need an additional search path for ROMs via an environment variable.

@pseudo-rnd-thoughts I'd probably try updating vcpkg to the latest version, they usually fix any major issues in a timely manner.

**Comment by pseudo-rnd-thoughts** [comment]
@JesseFarebro that makes sense, we will add an environment variable for overriding the rom path 

@JesseFarebro i don't understand vcpkg enough, could you make a pr for it?

**Comment by jjshoots** [comment]
We resorted to using an environment var instead of a kwarg. It's in #522

File: .github/workflows/ci.yml
Comment (pseudo-rnd-thoughts) [reply .github/workflows/ci.yml]: Fixed in #523

**Comment by pseudo-rnd-thoughts** [comment]
Thanks @JesseFarebro for your review, I'm going to merge this PR and cut a release in the next few days (hoping that the wheels ci works)

**Comment by Tobias-Fischer** [comment]
Would you be ok to open a PR @pseudo-rnd-thoughts? Or even better take over maintenance here?

**Pull request merged**
